---
title: "GWAS Workshop"
output: html_document
---
# Examine the distribution of the phenotype (IC50) in R

data <- read.table("D:/GWAS/GWAS_workshop/GS451_IC50.txt", header = TRUE)
head(data)
# Examine the distribution of the phenotype (IC~50~) in R
## To view the first few lines of the phenotype data:

## To view the distribution of the phenotype, 
## make a histogram:
  
hist(data$GS451_IC50,cex.lab=0.7,cex.main=0.7,cex.axis=0.7)


#Shapiro-Wilk test
shapiro.test(data$GS451_IC50)


# Transform the phenotype (IC~50~) in R
# Add the log2(IC50) phenotype to our data file in R
log2_GS451_IC50 <- log2(data$GS451_IC50)
head(log2_GS451_IC50)

#Shapiro-Wilk test on transformed data

shapiro.test(log2_GS451_IC50)


#make a histogram
hist(log2_GS451_IC50,cex.lab=0.7,cex.main=0.7,cex.axis=0.7)
data <- cbind(data, log2_GS451_IC50)
write.table(data, 
            file = "D:/GWAS/GWAS_workshop/GS451_IC50_log2.txt", 
            quote = FALSE, 
            row.names = FALSE)
head(data)


#Install Plink(use the Terminal)
 wget https://s3.amazonaws.com/plink1-assets/plink_linux_x86_64_20240804.zip


# Running the GWAS (Update the directory accordingly, Run the plink command in the terminal)
./plink --bfile /mnt/d/GWAS/GWAS_workshop/genotypes --maf 0.05 --linear --adjust --pheno /mnt/d/GWAS/GWAS_workshop/GS451_IC50_log2.txt --pheno-name log2_GS451_IC50 --out /mnt/d/GWAS/GWAS_workshop/GS451_L2_IC50_test


# Read in results sorted by CHR and BP
data <- read.table("D:/GWAS/GWAS_workshop/GS451_L2_IC50_test.assoc.linear", header = TRUE)
head(data)

# Source the qqman script
source('D:/GWAS/GWAS_workshop/qqman.r')

# Create Q-Q Plot and Manhattan Plot
png(filename = "D:/GWAS/GWAS_workshop/qqplot.png", res = 100)
 qq(data$P)
 dev.off()

png(filename = "D:/GWAS/GWAS_workshop/manplot.png", res = 100)
 manhattan(data)
 dev.off()

# First we must use PLINK to pull out the genotypes of rs7257475 in a format appropriate for R in the terminal
./plink --bfile genotypes --snp rs7257475 --recode A --out rs7257475_genotypes



# Make a boxplot of top SNP using R
gts <- read.table("D:/GWAS/GWAS_workshop/rs7257475_genotypes.raw", header = TRUE)
pts <- read.table("D:/GWAS/GWAS_workshop/GS451_IC50_log2.txt", header = TRUE)
boxplot(pts$log2_GS451_IC50 ~ gts$rs7257475_T, 
        ylab = "GS451 log2(IC50)", 
        xlab = "rs7257475 genotype")

